### ISSUE TITLE
Custom ElevenLabs Voices

### ISSUE THREAD

**Issue by Tybost** [Adri6336/gpt3-speech-to-text-chatbot#1]
Hey there! _Awesome work_ =)

- Could you add support for using Custom Cloned Elevenlabs voices? Thank You!
- It would also be cool to have an 'always listening' option to toggle (or say out loud)
- Ability to automatically select Elevenlabs as the default on launch _(I'm unable to switch back to Elevenlabs after restarting even with the voice command for some reason)_

My intention is to use this as an easy gateway for creating an A.I Virtual Content Creator like Neuro-Sama / or Nothing, Forever
https://www.youtube.com/watch?v=yn0iVOtr6FE&t=120s | https://youtu.be/jAWS3K_3jqQ?t=58

**Comment by Adri6336** [comment]
Heyo, thanks a mil! I'm super happy to hear that you like it :D. I'll def look into implementing your suggestions.

As for your third point, could you describe what your step-by-step process is? I tried to replicate it by starting with my Elevenlabs key, making the bot speak like a robot, then having it stop speaking like a robot. On my end the bot used the Elevenlabs TTS after the commands. 

If you happen to be using the keys.txt file, this may be a bug that I noticed and fixed. For some reason, my "(optional)" tag was interfering with the regex and preventing the bot from loading the Elevenlabs key at startup. I fixed it by moving the tag to the front of the key name and editing my regex, and have been able to start the bot with Elevenlabs as default since.

![keys](https://user-images.githubusercontent.com/64619524/219827335-9dd2a91d-0640-4064-96dd-0b7f5045ff58.png)

Downloading the most up-to-date version of the repo should have the bug-fix included (it's not yet part of a release). If you're using git, please be aware that updating will remove your keys from the file. To protect them from being overridden, make a copy first, update, then paste the keys into the new keys.txt file.

**Comment by Tybost** [comment]
This is embarrassing 😳 but I simply ran out of Quota with Elevenlabs (without realizing it) and so I upgraded my plan and it works fine again. 

I suppose there's one other (lower priority) thing to consider. It would be cool to have a built-in way to select which microphone input and cable output (much like how https://github.com/lugia19/speechToSpeechElevenLabs handles it) (Although windows built-in advanced sound options does work / but requires extra setup time :P)

**Comment by Adri6336** [comment]
😂 It's all good yo. Those character limits really do run out fast and catch you by surprise. Hopefully they increase the limits and speed up generation as they develop further. The ability to speak with a GPT bot that has an Elevenlabs voice in [real time](https://en.wikipedia.org/wiki/Real-time_computing) would be absolutely amazing!

I'm definitely planning to add more customization options, so setting the generation model and voice are on my to-do list. I don't have an immediate idea of how to go about selecting specific input / output cables, but I'll look into it and def check out the repo you linked more thoroughly. 

First though, I'm looking to add in token recycling. Based on my understanding of GPT-3, I'm pretty certain that a given conversation will fail if it goes on for too long (the prompt will accumulate far too many tokens). To get around this, I'm planning to add a feature where the most important gist of the conversation is retained while the back and forth messages are removed from the prompt. Once this is done, the bot should be generally reliable enough to work with over a longer period of time.

**Comment by Tybost** [comment]
ChatGPT API is out!

**Comment by Adri6336** [comment]
Just updated the bot to use ChatGPT by default! What a time to be alive :D
